hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java

```diff
@@ -572,8 +595,39 @@ private static boolean hiveEngineEnabled(TableMetadata metadata, Configuration c
         ConfigProperties.ENGINE_HIVE_ENABLED, TableProperties.ENGINE_HIVE_ENABLED_DEFAULT);
   }
 
+  /**
+   * Returns if the hive locking should be enabled on the table, or not.
+   *
+   * <p>The decision is made like this:
+   *
+   * <ol>
+   *   <li>Table property value {@link TableProperties#HIVE_LOCK_ENABLED}
+   *   <li>If the table property is not set then check the hive-site.xml property value {@link
+   *       ConfigProperties#LOCK_HIVE_ENABLED}
+   *   <li>If none of the above is enabled then use the default value {@link
+   *       TableProperties#HIVE_LOCK_ENABLED_DEFAULT}
+   * </ol>
+   *
+   * @param metadata Table metadata to use
+   * @param conf The hive configuration to use
+   * @return if the hive engine related values should be enabled or not
+   */
+  private static boolean hiveLockEnabled(TableMetadata metadata, Configuration conf) {
```

Comment:
Why do we want to override the table property using conf?  I thought all writes should use the hive lock enabled, or all writers should not.

Reply:
The table property could override the catalog property.
This way there is a graceful way to enable the config for every writer at the same time:
- Start by disabling the config on table level
- Enable the writers one-by-one based on the writers release process
- When you are sure that every writer uses the new code, then you can change the table config and all of the writers start using the new locking method at the same time

Reply:
That sounds ok if its a double-check, but the way the code here looks, if table config is set to NO_LOCK_ENABLED, it doesnt matter what the writer sets?   So looks like table config is all that matters.  Should we improve this check to make sure both writer/table config is NO_LOCK, as you descirbed?

Reply:
Actually  I was thinking more about it, even if we were to make this check return only if both are NO_LOCK, it still doesnt address the original concern that if one writer does not have NO_LOCK  but the table has NO_LOCK, then that writer will not switch to NO_LOCK and has chance to be corrupted by others that do.

So I see two options:

- all writers just follow table's NO_LOCK config
- writer could do a check, if table is set to NO_LOCK but they are not set to NO_LOCK, throw an exception and prevent the write.

Reply:
My thinking was somewhat different. 
I expect that we should highlight in the documentation the possible issues, and then with the configuration provide flexibility to use it correctly.

I expect that if someone does the effort of backporting the required changes, then they have one of the following:
1. They control everything and migrate with a big boom
2. Few problematic tables which they need to write without locks, but they do not want to touch the other tables.
3. They have a few tables which are written by old clients and most of the tables will be written by new clients

In the case of 2/3 there might be a situation where a particular client would like to write one table with locks and one table without locks, but there is a general rule which most of the clients would like to adhere.

Do I overcomplicate things? If we do not expect this kind of situation, then I would prefer a simpler/table level conf.

What do you think?
Thanks, Peter

Reply:
Sorry I'm trying but still not understanding how having a client level (catalog level) config, that is fallback to table config, allows case 2/3?  

To me as it is, it seems it causes more potential problem than solves (client says NO_LOCK but table does not have NO_LOCK).  Unless we want to do catalog config as a safety check (throw exception if catalog config is NO_LOCK but table does not have NO_LOCK), as I was mentioning before, which makes more sense to me.

Reply:
I think I finally understand our differences here.

- You think about Iceberg as a service, and the users as clients. In this view we have service level configurations (TableProperties), and client level configurations (CatalogProperties)
- I think about Iceberg as a library which provides different levels of configuration. The configuration provided by the developer (CatalogProperties), and global overrides which prevent configuration discrepancies between different users (TableProperties)

Does this make sense?
Am I right in understanding the different point of views?

Reply:
No I am not thinking of Iceberg like a service, but in my thought, indeed, the TableProperties are a single configuration persisted on the table, and the CatalogProperties is set by individual clients that can override.

My concern in the previous comments was, say we have one table, written by clients A and B.  Table property for lock is not set.

- Client A sets catalog property LOCK_HIVE_ENABLED = "false"
- Client B has catalog property LOCK_HIVE_ENABLED = "true"

This will make clients corrupt each other commit.   So was suggesting, either 

- Remove catalog config altogether
- Have Client A throw exception in this case (as I was mentioning, this may be better)

Does that make sense, or am I missing something?

Reply:
> Remove catalog config altogether

Then it is not possible to handle 2/3 use-cases (We either have to set the flag for every table, or rely on the default provided by the Iceberg source code - I think it is more useful for the user to provide an override possibility on catalog level too). I am open to debate the usefulness of this cases, but if we want to support them, then we need 2 levels of configuration.

> Have Client A throw exception in this case (as I was mentioning, this may be better)

We can throw an exception if the Table level and the Catalog level config is contradictory. In this case every table used by a Catalog instance should have the same locking configuration if it is set at least for one of the tables used.
If we decide on this then I agree with you that there is no need to have a Catalog level configuration, as it is just for throwing exceptions in some cases.


I am starting to feel that the main difference is:
- What do we consider more important:
    - Configuration flexibility: If the user wants to set all of the tables to use the new Locking mechanism, they should not need to go to every one of the tables and alter them one-by-one. A global catalog configuration should be enough in these cases. For me the question was how to handle when we would like to use a table differently than the general one and the table level config is proposed to do that.
    - Preventing wrong configurations: With the "only Table level config" solution we can be sure that if every writer uses which uses a code version which contains these changes will not have a writing conflict. BTW this can be archived even with the previous proposal if all of the tables are altered one-by-one to set the Table level config.

Reply:
OK I think I get the difference as well, if I can summarize, we have two choices

1.  catalog config = NO_LOCK => start writing to all tables with no_lock, unless table is set explicitly to LOCK.
2.  catalog config = NO_LOCK => start writing to tables with no_lock only if table set explicitly to NO_LOCK.

In both cases, we can achieve case 2/3, the issue is whether the user has to set the tables explicitly to exclude or include them in new system.

You are right that maybe Option 2 is too much of a burden for the user to manually mark all tables as NO_LOCK.  If all Iceberg clients are upgraded in the system, Option 2 may be a good check as then we are sure all of them will write NO_LOCK together, but we come back to the same problem if some clients are not upgraded.  (which is what case 2/3 is trying to solve).

Let me think about it, but I think you are right that we cannot do better in the mixed-version scenarios.